<launch>
  <!-- isrl robot model -->
  <include file="$(find isrl_urdf)/launch/isrl_bringup.launch"/>

  <!-- move base -->
  <!-- <include file="$(find isrl_navi)/launch/move_base.launch"/> -->
  
  <!-- Sync Node -->
  <!-- <node pkg="sync_package" type="sync_node" name="sync_node" output="screen"/> -->

  <arg name="database_path" default="/home/isrl-orin1/Documents/rtabmap_test.db"/>
  <arg name="localization" default="false"/>
  <arg name="with_camera" default="true"/>
  <arg name="wait_for_transform" default="0.2"/>
  <arg name="device_type_camera" default=""/>
  <arg name="serial_no_camera" default=""/>
  <arg name="camera" default="camera"/>
  <arg name="clip_distance" default="-2"/>
  <arg name="rviz" default="true"/>
  <arg name="rtabmapviz" default="true"/>

  <arg if="$(arg localization)" name="args" default=""/>
  <arg unless="$(arg localization)" name="args" default="--delete_db_on_start"/>

  <!-- Image transport node for decompressing images -->
  <!-- <node pkg="image_transport" type="republish" name="rgb_image_republish" args="compressed in:=/zed_node/rgb/image_rect_color raw out:=/zed_node/rgb/image_rect_color"/>
  <node pkg="image_transport" type="republish" name="depth_image_republish" args="compressedDepth in:=/zed_node/depth/depth_registered raw out:=/zed_node/depth/depth_registered"/> -->

  <!-- RTAB-Map Node -->
  <group ns="rtabmap">
    <node name="rtabmap" pkg="rtabmap_slam" type="rtabmap" output="screen" args="$(arg args)">

      <!-- ORB 기본 매개변수 -->
      <param name="ORB/MaxFeatures" type="int" value="500" />
      <param name="ORB/ScaleFactor" type="double" value="1.2" />
      <param name="ORB/NLevels" type="int" value="10" />
      <param name="ORB/EdgeThreshold" type="int" value="20" />
      <param name="ORB/FirstLevel" type="int" value="0" />
      <param name="ORB/WTA_K" type="int" value="4" />
      <!-- <param name="ORB/ScoreType" type="int" value="0" /> -->
      <param name="ORB/PatchSize" type="int" value="31" />
      <param name="ORB/FastThreshold" type="int" value="20" />
      <param name="ORB/GPU" type="string" value="true"/>

      <!-- GFTT 파라미터 -->
      <param name="GFTT/MinDistance" type="double" value="3.0" />
      <param name="GFTT/QualityLevel" type="double" value="0.02" />
      <param name="GFTT/BlockSize" type="int" value="3" />
      <param name="GFTT/UseHarrisDetector" type="bool" value="false" />
      <param name="GFTT/K" type="double" value="0.04" />

      <!-- feature matching type -->
      <param name="Vis/FeatureType" type="int" value="8"/>
      <param name="Kp/DetectorStrategy" type="int" value="8"/>

      <param name="database_path" type="string" value="$(arg database_path)"/>
      <param name="frame_id" type="string" value="base_footprint" />
      <param name="odom_frame_id" type="string" value="odom"/>
      <param name="wait_for_transform_duration" type="double" value="$(arg wait_for_transform)"/>
      <param name="queue_size" type="int" value="10"/>
      <param name="approx_sync" type="bool" value="true"/>
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_rgb" type="bool" value="true"/>
      <param name="subscribe_scan" type="bool" value="true"/>

      <!-- inputs -->
      <remap from="scan" to="/scan"/>
      <remap from="rgb/image" to="/zed_node/rgb/image_rect_color"/>
      <remap from="depth/image" to="/zed_node/depth/depth_registered"/>
      <remap from="rgb/camera_info" to="/zed_node/rgb/camera_info"/>
      
      <!-- output --> 
      <remap from="grid_map" to="/map"/>
      <remap from="proj_map" to="/proj_map"/>

      <!-- RTAB-Map's parameters -->
      <param name="RGBD/NeighborLinkRefining" type="string" value="true"/>
      <param name="RGBD/ProximityBySpace" type="string" value="true"/>
      <param name="RGBD/ProximityByTime" type="string" value="false"/>
      <param name="RGBD/ProximityPathMaxNeighbors" type="string" value="10"/>
      <param name="Reg/Strategy" type="string" value="2"/> <!-- 0=Visual, 1=ICP, 2=Visual+ICP -->
      <param name="Vis/MinInliers" type="int" value="10"/>
      <param name="RGBD/OptimizeFromGraphEnd" type="string" value="true"/>
      <param name="RGBD/OptimizeMaxError" type="double" value="1.0"/>
      <param name="Reg/Force3DoF" type="string" value="true"/>
      <param name="Grid/FromDepth" type="string" value="true"/>
      <param name="Mem/STMSize" type="double" value="30.0"/>
      <param name="RGBD/LocalRadius" type="double" value="7.0"/>
      <param name="Icp/CorrespondenceRatio" type="string" value="0.4"/>

      <!-- odom fix trying.. -->
      <param name="Odom/ResetCountdown" type="int" value="20" />

      <!-- Additional parameters for better obstacle detection -->
      <param name="Grid/RayTracing" type="string" value="true"/>
      <param name="Grid/3D" type="string" value="true"/>
      <param name="Grid/MaxGroundHeight" type="double" value="0.1"/>
      <param name="Grid/RangeMax" type="double" value="5.0"/>
      <param name="Grid/RangeMin" type="double" value="0.2"/>
      <param name="Grid/MaxObstacleHeight" type="double" value="0.8"/>

      <!-- localization mode -->
      <param if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false"/>
      <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true"/>
      <param name="Mem/InitWMWithAllNodes" type="string" value="$(arg localization)"/>
    </node>

    <!-- visualization with rtabmapviz -->
    <node if="$(arg rtabmapviz)" pkg="rtabmap_viz" type="rtabmap_viz" name="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
      <param name="subscribe_scan" type="bool" value="true"/>
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_rgb" type="bool" value="true"/>
      <param name="subscribe_odom" type="bool" value="true"/>
      <param name="frame_id" type="string" value="base_footprint"/>
      <param name="approx_sync" type="bool" value="true"/>

      <remap from="odom" to="/odom"/>
      <remap from="scan" to="/scan"/>
      <remap from="rgb/image" to="/zed_node/rgb/image_rect_color"/>
      <remap from="depth/image" to="/zed_node/depth/depth_registered"/>
      <remap from="rgb/camera_info" to="/zed_node/rgb/camera_info"/>
    </node>
  </group>

  <!-- rviz -->
  <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find isrl_slam)/rviz/rtabslam.rviz" output="screen"/>
</launch>
